### PR TITLE
Add support for pnpm workspaces as single derivation

### PR DESCRIPTION
**Pull request by iteratee**
Allow a caller to specify a workspace and a list of components rather than a single source directory. This allows for the building of pnpm monorepo projects with dependency links between the various projects and a single pnpm-lock.yaml for the whole workspace. Requires that the script has the same name in all components.

The location of all of the per-component package.json files is overridable. A workspace project can still specify the location of the root package.json via packageJSON, as well as the components' package.json files via componentPackageJSONs. The list defaults to c/package.json for c in components.

Similarly, the list of distDirs for the components is overridable. For a workspace, the distDirs are all subdirs of $out rather than dist becoming $out. The default list is c/dist for c in components. Allowing for override handles the case where one component is built by a different tool like next and produces a different directory like ".next"

NB: Any "link:" dependencies will need to be recreated as a preBuild step. pnpm will create the symlinks during install, and nix removes them because when the node_modules derivation is built, they are dangling.

Attempts to leave the non-workspace pnpm package support intact.

### CODE REVIEW

**Comment by iteratee** [comment]
Feel free to ask me anything about the changes, even if you feel like it's a dumb question or a nix noob question. I had to learn a lot about nix and flakes just to be able to write this. I wouldn't expect anyone who hadn't worked a lot with nix or written something like this to be able to review without asking a fair number of questions.

File: derivation.nix [hunk 1]
```diff
@@ -15,15 +15,23 @@ let
...
   mkPnpmPackage =
-    { src
+    { workspace ? null
+    , components ? []
+    , src ? if (workspace != null && components != []) then workspace else null
     , packageJSON ? src + "/package.json"
-    , pnpmLockYaml ? src + "/pnpm-lock.yaml"
+    , componentPackageJSONs ? map (c: {
+        name = "${c}/package.json";
+        value = src + "/${c}" + "/package.json";
```
Comment (iteratee) [inline derivation.nix]: Can combine strings here to remove one of the `+` operators.

File: derivation.nix [hunk 2]
```diff
@@ -15,15 +15,23 @@ let
...
+      }) components
+    , pnpmLockYaml ? (if workspace == null then src else workspace) + "/pnpm-lock.yaml"
```
Comment (iteratee) [inline derivation.nix]: It would be simpler to not attempt to support the case where `workspace != src`. If I make that assumption, I can drop this diff.

File: derivation.nix [hunk 3]
```diff
@@ -37,17 +45,40 @@ in
...
+        (if pnpmWorkspaceYaml == null
+          then []
+          else [
+            {name = "pnpm-workspace.yaml"; value = pnpmWorkspaceYaml;}
+          ]
+        ) ++ extraNodeModuleSources;
+      copyLink =
+        if copyNodeModules
+          then "cp -r"
+          else "ln -s";
```
Comment (iteratee) [inline derivation.nix]: All of the definitions here are just helpers to simplify the changes below. So is `isWorkspace`.

File: derivation.nix [hunk 4]
```diff
@@ -37,17 +45,40 @@ in
...
+      forEachComponent = f: forEachConcat f components;
     in
     stdenv.mkDerivation (
       recursiveUpdate
         (rec {
           inherit src name nativeBuildInputs;
 
+          postUnpack = ''
+            ${optionalString (pnpmWorkspaceYaml != null) ''
+              cp -v ${pnpmWorkspaceYaml} pnpm-workspace.yaml
```
Comment (iteratee) [inline derivation.nix]: can drop the `-v` (verbose) flag. Debugging that is no longer needed.

File: derivation.nix [hunk 5]
```diff
@@ -70,15 +107,35 @@ in
           buildPhase = ''
             runHook preBuild
 
-            pnpm run ${script}
+            ${if isWorkspace
+              then ''
+                ${forEachComponent (component: ''
+                  (cd "${component}" && pnpm run ${script})
```
Comment (iteratee) [inline derivation.nix]: Looking at this loop, and the loop above, it may be better to define a computed components that is `["."]` for the non-workspace case. This would remove some of the `if isWorkspace` tests.

File: derivation.nix [hunk 6]
```diff
@@ -70,15 +107,35 @@ in
...
+            }
 
             runHook postBuild
           '';
 
           installPhase = ''
             runHook preInstall
 
-            ${if distDir == "." then "cp -r" else "mv"} ${distDir} $out
+            ${if isWorkspace
```
Comment (iteratee) [inline derivation.nix]: It would not remove this test, because the behavior is necessarily different from a workspace with a single component.

File: derivation.nix [hunk 7]
```diff
@@ -112,18 +169,21 @@ in
                 inherit nativeBuildInputs;
 
                 unpackPhase = concatStringsSep "\n"
-                  (
+                  ( [ # components is an empty list for non workspace builds
```
Comment (iteratee) [inline derivation.nix:199]: The comment would need to change here, but `mkdir -p .` is a no-op, so having a non-empty components list wouldn't hurt in this case.

File: derivation.nix [hunk 8]
```diff
@@ -112,18 +169,21 @@ in
...
+                      (forEachComponent (component: ''
+                      mkdir -p "${component}"
+                    '')) ] ++
                     map
                       (v:
                         let
                           nv = if isAttrs v then v else { name = "."; value = v; };
                         in
-                        "cp -vr ${nv.value} ${nv.name}"
+                        "cp -vr \"${nv.value}\" \"${nv.name}\""
```
Comment (iteratee) [inline derivation.nix:208]: The quotes here could go in a separate commit. I think they are the only change that would be just a simple improvement to the existing code. The quotes are necessary now, because the component names are now part of the paths here, and are user-supplied.

**Comment by iteratee** [comment]
I should update the README to document the new features as well. That can be a separate commit after this one that adds the functionality. I don't think that should block our review of this code, but should be added before creating a PR for upstream.

File: derivation.nix (hunk 1 above)
Comment (iteratee) [reply derivation.nix]: Done.

File: derivation.nix (hunk 5 above)
Comment (iteratee) [reply derivation.nix]: I decided against this. Shorter code is not worth conflating components and non-components builds. I did however pre-compute all of the things that need to be different for a workspace build, many of them into lists, and then the derivation just loops over those lists.

File: derivation.nix (hunk 2 above)
Comment (iteratee) [reply derivation.nix]: Done.

File: derivation.nix (hunk 4 above)
Comment (iteratee) [reply derivation.nix]: Done.

File: derivation.nix (hunk 6 above)
Comment (iteratee) [reply derivation.nix]: I instead made a different flag that controls this specific behavior. I think this is cleaner.

File: derivation.nix (hunk 7 above)
Comment (iteratee) [reply derivation.nix:199]: Comment stays, as components is empty list for non-workspace build.

File: derivation.nix [hunk 9]
```diff
@@ -26,6 +26,8 @@ in
     , distDir ? "dist"
     , installInPlace ? false
     , installEnv ? { }
+    , installAndBuildEnv ? { }
```
Comment (RocketPuppy) [inline derivation.nix]: My preference would be to not have this shared env argument. If the user is going to de-duplicate their ENV attrsets to use them here they could just as easily do `sharedEnv // installEnv` and `sharedEnv // buildEnv` on the call side.

Comment (iteratee) [reply derivation.nix]: Fair point. That does simplify the interface.

File: derivation.nix [hunk 10]
```diff
@@ -39,17 +47,66 @@ in
...
+        ) ++ extraNodeModuleSources;
+      # Computed values that loop over something
+      nodeModulesDirs =
+        if isWorkspace then
+          ["node_modules"] ++ (map (c: "${c}/node_modules") components)
+        else ["node_modules"];
+      buildScripts =
+        if isWorkspace then ''
+          ${forEachComponent (component: ''
+            (cd "${component}" && pnpm run ${script})
```
Comment (RocketPuppy) [inline derivation.nix]: Can `pnpm --recursive run ${script}` work without needing the component list?

**Comment by RocketPuppy** [comment]
I think the complicated pieces of this aren't really Nix related but related to how Turbo/PNPM interact with workspaces. I'm not really sure how to fix that.

File: derivation.nix (hunk 10 above)
Comment (iteratee) [reply derivation.nix]: We need the components list to knowh which `node_modules` directories we need and for install to know what to copy.

I might be able to parse the pnpm-workspace.yaml file and remove the ability to specify the components.

As long as we accept the components list, it makes more sense to just build each component. There is no guarantee the caller gave us all the components. I might be able to rig something up with filters, but that feels more fragile than just accepting directories as arguments and using them as directories.

File: derivation.nix (hunk 9 above)
Comment (iteratee) [reply derivation.nix]: Done.

**Review by RocketPuppy — APPROVED** [review]
Tentatively approving this. I don't have any other feedback about the nix stuff and the actual JS build process I don't know enough about to be constructive.

File: derivation.nix (hunk 10 above)
Comment (iteratee) [reply derivation.nix]: It turned out that using a filter list was less fragile. A component may be needed without having a particular script.

So I went with a filtered `pnpm run --recursive`.

File: derivation.nix [hunk 11]
```diff
@@ -15,15 +16,23 @@ let
...
   mkPnpmPackage =
-    { src
+    { workspace ? null
+    , components ? []
+    , src ? if (workspace != null && components != []) then workspace else null
     , packageJSON ? src + "/package.json"
+    , componentPackageJSONs ? map (c: {
+        name = "${c}/package.json";
+        value = src + "/${c}/package.json";
+      }) components
```
Comment (cbarber) [inline derivation.nix:26]: Should log/fail here if these paths do not exist? My thinking is someone checks out marketplace and set ups direnv/nix prior to `git submodule update` which would mean `nucleus/packages.json` is missing. Not worthwhile if this would be tricky since we won't have the submodule for much longer.

**Review by cbarber — APPROVED** [review]
Approving leaving the missing package.json log/failure to your discretion.
